lib/galaxy/util/watcher.py

```diff
@@ -0,0 +1,169 @@
+# module should probably be updated to use this where possible
+
+from __future__ import absolute_import
+
+import logging
+import os.path
+import time
+
+from six.moves import filter
+
+try:
+    from watchdog.events import FileSystemEventHandler
+    from watchdog.observers import Observer
+    from watchdog.observers.polling import PollingObserver
+    can_watch = True
+except ImportError:
+    Observer = None
+    FileSystemEventHandler = object
+    PollingObserver = None
+    can_watch = False
+
+from galaxy.util.hash_util import md5_hash_file
+from galaxy.web.stack import register_postfork_function
+
+log = logging.getLogger(__name__)
+
+
+def get_observer_class(config_name, config_value, default, monitor_what_str):
+    """
+    """
+    config_value = config_value or default
+    config_value = str(config_value).lower()
+    if config_value in ("true", "yes", "on", "auto"):
+        expect_observer = True
+        observer_class = Observer
+    elif config_value == "polling":
+        expect_observer = True
+        observer_class = PollingObserver
+    elif config_value in ('false', 'no', 'off'):
+        expect_observer = False
+        observer_class = None
+    else:
+        message = "Unrecognized value for %s config option: %s" % (config_name, config_value)
+        raise Exception(message)
+
+    if expect_observer and observer_class is None:
+        message = "Watchdog library unavailable, cannot monitor %s." % monitor_what_str
+        if config_value == "auto":
+            log.info(message)
+        else:
+            raise Exception(message)
+
+    return observer_class
+
+
+def get_watcher(config, config_name, default="False", monitor_what_str=None, watcher_class=None,
+                event_handler_class=None, **kwargs):
+    config_value = getattr(config, config_name, None)
+    observer_class = get_observer_class(config_name, config_value, default=default, monitor_what_str=monitor_what_str)
+    if observer_class is not None:
+        watcher_class = watcher_class or Watcher
+        event_handler_class = event_handler_class or EventHandler
+        return watcher_class(observer_class, event_handler_class, **kwargs)
+    else:
+        return NullWatcher()
+
+
+class Watcher(object):
+
+    def __init__(self, observer_class, event_handler_class, **kwargs):
+        self.monitored_dirs = {}
+        self.path_hash = {}
+        self.file_callbacks = {}
+        self.dir_callbacks = {}
+        self.ignore_extensions = {}
+        self.observer = observer_class()
+        self.event_handler = event_handler_class(self)
+        self.start()
+
+    def start(self):
+        register_postfork_function(self.observer.start)
+
+    def shutdown(self):
+        self.observer.stop()
+        self.observer.join()
+
+    def monitor(self, dir, recursive=False):
+        self.observer.schedule(self.event_handler, dir, recursive=recursive)
+
+    def watch_file(self, file, callback=None):
+        file = os.path.abspath(file)
```

Comment:
do we maybe want to change `file` to `path` or `fpath`, `file_path`  or some other non-builtin ? Same with `dir`